mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeansModel.scala

```diff
@@ -225,13 +227,14 @@ object BisectingKMeansModel extends Loader[BisectingKMeansModel] {
       assert(formatVersion == thisFormatVersion)
       val rootId = (metadata \ "rootId").extract[Int]
       val distanceMeasure = (metadata \ "distanceMeasure").extract[String]
+      val trainingCost = (metadata \ "trainingCost").extract[Double]
```

Comment:
hmm, can this read old model from previous version?

Reply:
- Could you avoid modifying loading model code in "mllib" package, but modifying code in "ml" package, i.e., the class `ml.clustering.BisectingKMeansModel.BisectingKMeansModelReader`, you can reference the `KMeans` code: `ml.clustering.KMeansModel.KMeansModelReader`.

- And, +1 with @viirya mentioned, we should keep model loading compatibility, add a version check (when >= 2.4) then we load "training cost" . Note that add these in `ml.clustering.BisectingKMeansModel.BisectingKMeansModelReader`.

- And, could you also add version check (when >= 2.4) then we load "training cost" into `ml.clustering.KMeansModel.KMeansModelReader` ?

Reply:
Do other models have this problem? I was told that this change just follows what we did for other models before.

Reply:
Thank you all for the comments and sorry for the late answer. Just a couple of notes on your comments @WeichenXu123  (I may be missing something, so please correct me if I am wrong):

 - I checked the `ml.clustering.KMeansModel.KMeansModelReader` and it doesn't store anything related to the summary. Summary is not recovered after load of the model, so I don't see any reason why we should modify the read/load of `ml.clustering.BisectingKMeansModel.BisectingKMeansModelReader`;
 - this model can read from previous versions, since this is version "2.0", which was introduced for Spark 2.4; for previous versions, we read/write version "1.0"; the version check method for versioning is used only for the `ml` package, not in `mllib` where we have this versioning approach;

> I was told that this change just follows what we did for other models before.

@cloud-fan Yes, let me link the PR for KMeans doing the same, which is: https://github.com/apache/spark/pull/20629.

Just a final comment which I hope clarifies which is the source of the confusion here and the reason of the above comments by @viirya and @WeichenXu123: `trainingCost` here is a member of the `summary`, not a parameter of the model for the `ml.clustering.BisectingKMeansModel`. Instead, it is a member of the model for `mllib.clustering.BisectingKMeansModel` (we have no summary notion there...). So storing it for `mllib` is needed in order for the model read after persisting it being the same of the original one (I think it doesn't pass UTs otherwise). Storing it for the `ml`, instead, it is not needed, because the summary is not persisted. If we want to persist also the summary for `ml` package I think we should best create a separate JIRA and PR for it.

Hope this clarifies (sorry for being so verbose).

Reply:
> this model can read from previous versions, since this is version "2.0", which was introduced for Spark 2.4; for previous versions, we read/write version "1.0"; the version check method for versioning is used only for the ml package, not in mllib where we have this versioning approach;

I meant that can it read old model from previous versions, not that this model can read from previous versions.

In other words, when reading a previous model without "trainingCost" in metadata, can this line work well?

```scala
val trainingCost = (metadata \ "trainingCost").extract[Double]
```

Reply:
The answer is yes, because of versioning:
 - all the versions of Spark released so far, use the versions `"1.0"`, which is loaded/written with the `SaveLoadV1_0` class (see https://github.com/apache/spark/pull/22764/files#diff-458b0e49bfc4a4dd993ea6cb66af4b4dR125);
 - there can't be a model with version `"2.0"` without `trainingCost`, because `SaveLoadV2_0` was introduced in a previous commit which lands only on 2.4 and has not yet been released.

Reply:
Ok. I see. `SaveLoadV2_0` is confused term as I think it is for reading models since Spark 2.0.

Reply:
@mgaido91 
What about this way ?
In `ml` reader, we can load `trainingCost` and then construct `mllib.clustering.BisectingKMeansModel` and pass the `trainingCost` argument and then construct `ml.clustering.BisectingKMeansModel` to wrap that `mllib` model.
** Then in `ml` reader we can check the spark major/minor version to keep backwards compatibility, this is a more important thing. **
** In `mllib` loader we can ignore loading `trainingCost`, because `mllib` is deprecate, we can ignore adding new features, but we cannot breaking backwards compatibility (your change breaking backwards compatibility). **

Reply:
@WeichenXu123 I don't see (and think) this change breaks backwards compatibility for `mllib`. You can still load a model stored by previous versions, the only difference is that it won't have a valid `trainingCost` (but this just means we lack a new feature, all operations/data available before is still there, so backward compatibility is ensured). The same approach was used also for `KMeans`, so if we want to change approach here, we might have to do it also for `KMeans`.

Anyway, regarding your suggestion, since the summary is not recomputed, there is no need at all to store this for `ml`: a loaded model doesn't contain the summary, so doesn't contain the `trainingCost` (same for `KMeans`). We could/should rather ignore at all storing the value, it changes nothing on `ml` side.

On the other side, if we want it to be stored and re-loaded on `ml` side, we should start persisting the summary too there, but as I mentioned before, I think this is a different and bigger change, which should involve all the summary/models and not only `BisectingKMeans`.

Reply:
@mgaido91 
(I haven't test this, so correct me if I am wrong).
> I don't see (and think) this change breaks backwards compatibility for mllib.

I am suspicious of this line:
```
val trainingCost = (metadata \ "trainingCost").extract[Double]
```
When loading an old version spark saved `BisectingKMeansModel`, because it do not contain "trainingCost" info, I guess this line will throw error. (Otherwise what will it return ?)

Reply:
yes @WeichenXu123 , you're right, that line is a bug. Thanks for noticing it. Anyway, that is going to be addressed in another PR and it is not (strictly) related to this one. The other option, as I mentioned, is that if we agree that this doesn't need to be restored after model persistence, we can just ignore it in save/load.

Reply:
OK. After https://github.com/apache/spark/pull/22790 merged, I think this PR can work.